File: BaekJoon/Java/3649_로봇프로젝트_TwoPointer.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

// 로봇프로젝트 3649 BinarySearch
// 주어진 레고 조각들 중 두 조각의
// 합이 x 와 같은지 판별하는 문제
// O( nlogn )


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String str;

        while ((str = br.readLine()) != null) {
            int x = Integer.parseInt(str) * 10_000_000;
            int n = Integer.parseInt(br.readLine());
            int[] legos = new int[n];

            for (int i = 0; i < n; ++i) {
                legos[i] = Integer.parseInt(br.readLine());
            }

            // 오름차순
            Arrays.sort(legos);

            int l = 0, r = n - 1;
            boolean found = false;

            while (l < r) {
                int sum = legos[l] + legos[r];

                if (sum < x) {
                    ++l;
                } else if (sum > x) {
                    --r;
                } else {
                    System.out.println("yes " + legos[l] + " " + legos[r]);
                    found = true;
                    break;
                }
            }

            if (!found) System.out.println("danger");
        }
    }
}

File: BaekJoon/3649_로봇프로젝트_TwoPointer.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// 로봇프로젝트 3649 TwoPointer
// 주어진 레고 조각들 중 두 조각의
// 합이 x 와 같은지 판별하는 문제


int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int x;

    while (cin >> x)
    {
        x *= 10'000'000; // cm -> nm

        int n;
        cin >> n;

        vector<int> legos(n);
        for (int i = 0; i < n; ++i)
        {
            cin >> legos[i];
        }

        // 오름차순
        sort(legos.begin(), legos.end());

        int l{ 0 }, r{ n - 1 }; // 투 포인터
        int sum{ 0 };

        while (l < r)
        {
            // 두 레고 조각의 합
            sum = legos[l] + legos[r];

            if (sum < x) ++l;
            else if (sum > x) --r;
            else break;
        }

        // 구멍을 막을 수 있다면
        if (sum == x)
        {
            cout << "yes " << legos[l] << " " << legos[r] << "\n";
        }
        else
        {
            cout << "danger\n";
        }
    }


    return 0;
}
